_posts/2019/2019-09-22-a-rails-active-model-validator-for-stripe-ids.md

```diff
@@ -0,0 +1,77 @@
+---
+title: "A Active Model validator for Stripe Ids"
+description: "A decent example"
+layout: article
+category: ruby
+image:
+  base: '2019/a-rails-active-model-validator-for-stripe-ids'
+  alt: 'Zebra stripes'
+  credit: David Clarke
+  source: "https://unsplash.com/photos/sN6d60TySV0"
+
+---
+
+In [my article on using validators](/ruby/use-a-custom-validator), the example was pretty simple.
+
+This is an example of real production use of a more complex validator extracted from the billing code in [AnswerThePublic](https://answerthepublic.com). We use [Stripe](https://stripe.com) exclusively for the billing in our applications.
+
+
+## Instead of...
+
+...using a regex in `validates` when setting a format for the `stripe_id`:
+
+```ruby
+class Subscription < ActiveRecord
+  validates :stripe_id, format: { with: /\Asub\_[A-Za-z0-9]{8,}\z/ }
+end
+```
+
+
+## Use...
+
+...a validator class to ensure the format of a Stripe `id`...
+
+```ruby
+# app/validators/stripe_validator.rb
+class StripeValidator < ActiveModel::EachValidator
+  VALID_PREFIXES = %w[ch cus in plan sub txr].freeze
+
+  def validate_each(record, attribute, value)
+    record.errors[attribute] << (options[:message] || 'is not a valid Stripe ID') unless regex(options[:prefix]).match?(value)
+  end
+
+  def check_validity!
+    raise ArgumentError, 'Requires :prefix' unless options.include?(:prefix)
+    raise ArgumentError, 'Invalid :prefix supplied' unless VALID_PREFIXES.include?(options[:prefix].to_s)
+  end
+
+  private
+
+  def regex(value)
+    /\A#{value}\_[A-Za-z0-9]{8,}\z/
+  end
+end
+```
+
+...then used in models like so:
+
+```ruby
+class Subscription < ActiveRecord
+  validates :stripe_id, stripe: { prefix: :sub }
+end
+
+class Invoice < ActiveRecord
+  validates :stripe_id, stripe: { prefix: :in }
+end
+```
+
+## Why?
+
+We’ve built a validator that we reuse across multiple models in our own domain, that have records on Stripe accessible via their API.
+
+Hopefully this shows where having extra structure in your code can improve the clarity of understand when you, or another developer, come back to it.
```

Comment:
*or understanding?

Reply:
At this point I'm like: huh? So what is this ep of ORT showing me?

Reply:
Okay — I read this again, and I now get it, but I don't think it flowed so well. Some of my corrections will help, I reckon. But I think there needs a little more story at the beginning of what Stripe IDs are why it's important you have validators for them.

And maybe explain that you're giving a more specific Stripe example because you reckon most people will come across it at some point, not just: we use it all the time so here's some example code for you.